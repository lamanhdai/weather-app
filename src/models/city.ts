export interface City {
  title: string
  woeid: string
}