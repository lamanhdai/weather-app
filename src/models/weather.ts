export interface Weather {
  id: string
  weather_state_name: string
  weather_state_abbr: string
  wind_direction_compass: string
  applicable_date: string
  min_temp: number
  max_temp: number
  the_temp: number
  wind_speed: number
  wind_direction: number
  air_pressure: number
  humidity: number
  visibility: number
  predictability: string
}

export interface WeatherArea {
  weatherList: Weather[],
  city: string,
  woeid: string
}