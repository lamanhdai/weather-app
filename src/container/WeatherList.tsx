import React from 'react';
import {Container, Row, Col, Spinner} from 'react-bootstrap';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { InView } from 'react-intersection-observer';

import {WeatherArea} from 'models/weather';
import WeatherInfo from 'components/WeatherInfo';
import Search from 'components/Search';
import {WeatherState} from 'store/reducers';
import {AppActions} from 'store/actions.type';
import {getWeatherRequest} from 'store/actions.creator';


interface WeatherListProps {
  weatherListByCity: WeatherArea[]
  err: string
  onSearch: (keyword: string) => void
  loading: boolean
}

export function WeatherList (props: WeatherListProps) {
  const {onSearch, weatherListByCity, err, loading} = props;
  const renderWeather = (list: WeatherArea[]) => {
    return list&&list.length?list.map(weatherArea => (
      <InView key={weatherArea.woeid}>
        {({ ref }) => (
          <div className="mt-5" ref={ref}>
            <h2>City: {weatherArea.city}</h2>
            <Row className="mt-2" >
              {
                weatherArea.weatherList.map(weather => (
                  <Col key={weather.id}><WeatherInfo info={weather} /></Col>
                ))
              }
            </Row>
          </div>
        )}
        </InView>
      )):<div className="text-center mt-5">Oops ! no result</div>
    
  }

  const renderSpinner = () => {
    return <div className="text-center mt-5">
      <Spinner animation="border" role="status">
        <span className="visually-hidden">Loading...</span>
      </Spinner>
    </div>
  }
  return (
    <Container className="mt-5">
        <Search onSearch={onSearch} />
        {
          loading?
          renderSpinner()
          :
          renderWeather(weatherListByCity)
        }
        {
          err&&<div>{err}</div>
        }
    </Container>
  )
}

function mapStateToProp(state: WeatherState) {
  const {weatherListByCity, err, loading} = state;
  return {
    weatherListByCity,
    err,
    loading
  }
}

function mapDispatchToProp(dispatch: Dispatch<AppActions>) {
  return {
    onSearch: (keyword: string) => {
      dispatch(getWeatherRequest(keyword))
    }
  }
}

export default connect(mapStateToProp, mapDispatchToProp)(WeatherList);