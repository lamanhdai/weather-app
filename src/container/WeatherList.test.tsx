import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';

import {WeatherArea} from 'models/weather';
import configureStore from 'redux-mock-store';

configure({adapter: new Adapter()});

import {WeatherList, default as ConnectWeatherList} from './WeatherList';

describe('WeatherList', () => {
  const props = {
    weatherListByCity: [],
    err: '',
    onSearch: () => {},
    loading: false
  }

  it('render weather list no store', () => {
    const WeatherListElment = renderer
      .create(
          <WeatherList {...props} />
      )
      .toJSON();

    expect(WeatherListElment).toMatchSnapshot();
  });

  it('render weather list with store', () => {
    const mockStore = configureStore([]);
    const store = mockStore({
      weatherListByCity: [],
      err: '',
      loading: false
    });

    const WeatherListElment = renderer
      .create(
        <Provider store={store}>
          <ConnectWeatherList {...props} />
        </Provider>
      )
      .toJSON();

    expect(WeatherListElment).toMatchSnapshot();
  })
});