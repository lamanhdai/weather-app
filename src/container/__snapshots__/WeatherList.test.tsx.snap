// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherList render weather list no store 1`] = `
<div
  className="mt-5 container"
>
  <div
    className="mt-5"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="form-group form-group"
      >
        <input
          aria-describedby="search"
          className="form-control"
          placeholder="Search a location"
          required={true}
          role="search"
          type="text"
        />
      </div>
    </form>
  </div>
  <div
    className="text-center mt-5"
  >
    Oops ! no result
  </div>
  
</div>
`;

exports[`WeatherList render weather list with store 1`] = `
<div
  className="mt-5 container"
>
  <div
    className="mt-5"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="form-group form-group"
      >
        <input
          aria-describedby="search"
          className="form-control"
          placeholder="Search a location"
          required={true}
          role="search"
          type="text"
        />
      </div>
    </form>
  </div>
  <div
    className="text-center mt-5"
  >
    Oops ! no result
  </div>
  
</div>
`;
